### ISSUE TITLE
Debug points failing to bind in body of "[ for x in xs -> body ]"

### ISSUE THREAD

**Issue by dsyme** [dotnet/fsharp#13504]
See https://github.com/dotnet/fsharp/pull/12432#issuecomment-1156014831

> It's a good question - do you have a more specific repro of this problem please?

Just tested with the following demo:

```fsharp
[<EntryPoint>]
let main args =

    let test1 () =
        [ 3; 2; 1 ]
        |> List.map (fun x -> x + 10)
        |> List.sort

    let test2 = [
        for x in test1() do
            let xx = x * x
            printf "test"
            xx
        ]

    let test3 = [
        for x in test1() ->
            let xx = x * x
            printf "test"
            xx
        ]

    printf "The end"
```
![AaWk1DzlQL](https://user-images.githubusercontent.com/1535823/173750497-80c67368-1bd6-4dfc-ba4b-607bcf21bfc0.gif)

Notice that:
- I can add breakpoints in both `test2` and `test3` before starting the debugger (F5)
- When debugging, only the breakpoints in `test2` stop the debugger
- If I remove/re-add a breakpoint, it works in `test2`, it fails with the message "The breakpoint failed to bind." in `test3`

I'm using Visual Studio Community 2022 (64-bit) - Version 17.1.0, and it's a .NET 6.0 project.

**Comment by ntovas** [comment]
Can't reproduce in VS: `Community`/`Enterprise` 2022 (64-bit) - Version: `17.4.0 preview 2.1`/`latest main branch` with either `.net 6.0`/`.net 7.0 rc1`

**Comment by brianrourkeboll** [comment]
I ran into this and other debug point oddities in collection expressions in https://github.com/dotnet/fsharp/pull/17067#issuecomment-2085624187 — in fact, although the repro code above is now optimized to a fast `while`-loop, I kept the debug points exactly the same as before, which means that this still reproduces.

(It also reproduces if you write `for x in Seq.ofList (test1()) do` so that the optimization doesn't apply.)

It would be nice if there were a way to write tests for debug point locations. If that is already done somewhere, I haven't seen it.

**Comment by kirisky** [comment]
Is anyone working on this? If not, I'd like to give it a try.

**Comment by brianrourkeboll** [comment]
> The bug is still there. Is anyone working on this? If not, I'd like to give it a try.

I don't think anyone is looking at this currently, no.

Ideally we would:

1. Document how debugging _should_ behave in each scenario (`for`-`to`, `for`-`in` over ranges, lists, arrays, sequences).
2. Write tests asserting that behavior.
3. Update the codegen where needed to comply with that behavior.

You can see some examples of the differing behaviors in https://github.com/dotnet/fsharp/pull/17067#issuecomment-2085624187.

**Comment by kirisky** [comment]
Thanks, Brian! Will take a look!

**Comment by brianrourkeboll** [comment]
Depending on how far you want to go in trying to make things consistent, you may need to make changes in places like [TypedTreeOps.fs](https://github.com/dotnet/fsharp/blob/08a0649d0eb4d49c2a1277f2595cbc9804ba58c2/src/Compiler/Optimize/LowerComputedCollections.fs#L376-L439) as well as LowerComputedCollections.fs, and maybe others.

**Comment by kirisky** [comment]
Got it! I'll keep an eye on those. Really appreciate the heads-up, Brian!
